### PR TITLE
Reduce read contention in Raft FSM

### PR DESCRIPTION
**Pull request by manadart**
This is a simple mechanical change to use `sync.RWMutex` instead of `sync.Mutex` in our Raft FSM implementation.

We only need normal locks for mutating; I.e. `Apply` and `Restore`. Everything else is read-lock protected allowing concurrent reads.

## QA steps

Run test suite with `-race` flag.

### CODE REVIEW

**Comment by manadart** [comment]
!!build!!

**Review by SimonRichardson — APPROVED** [review]

File: core/raftlease/fsm.go [hunk 1]
```diff
@@ -267,20 +267,20 @@ func (f *FSM) filteredLeases(getLocalTime func() time.Time, keys []lease.Key) ma
...
-	f.mu.Lock()
+	f.mu.RLock()
 	localTime := getLocalTime()
 	for _, entries := range f.groups {
 		for key, entry := range entries {
 			results[key] = f.infoFromEntry(localTime, key, entry)
 		}
 	}
-	f.mu.Unlock()
+	f.mu.RUnlock()
```
Comment (juanmanuel-tirado) [inline core/raftlease/fsm.go:283]: `RUnlock`should be invoked using `defer` to maintain consistency with the other pieces of code.

File: core/raftlease/fsm.go [hunk 2]
```diff
@@ -324,14 +324,14 @@ func (f *FSM) LeaseGroup(getLocalTime func() time.Time, namespace, modelUUID str
...
-	f.mu.Lock()
+	f.mu.RLock()
 	pinned := make(map[lease.Key][]string)
 	for key, entities := range f.pinned {
 		if !entities.IsEmpty() {
 			pinned[key] = entities.SortedValues()
 		}
 	}
-	f.mu.Unlock()
+	f.mu.RUnlock()
```
Comment (juanmanuel-tirado) [inline core/raftlease/fsm.go:334]: Better with `defer`

File: core/raftlease/fsm.go [hunk 3]
```diff
@@ -442,7 +442,7 @@ func (f *FSM) Snapshot() (raft.FSMSnapshot, error) {
 		}] = entities.SortedValues()
 	}
 
-	f.mu.Unlock()
+	f.mu.RUnlock()
```
Comment (juanmanuel-tirado) [inline core/raftlease/fsm.go:445]: Better with `defer`

File: core/raftlease/fsm.go (hunk 1 above)
Comment (SimonRichardson) [reply core/raftlease/fsm.go:283]: defer was slower at one point, which was probably why it was written like this. In reality that's gone now, so probably worth refactoring for readability.

Comment (juanmanuel-tirado) [reply core/raftlease/fsm.go:283]: I add the Go notice about the defer improvement for the sake of completeness https://golang.org/doc/go1.14#runtime

**Comment by manadart** [comment]
$$merge$$

**Comment by manadart** [comment]
[Use RWMutex in FSM](https://trello.com/c/kGHvx7qc/3867-use-rwmutex-in-fsm)

**Pull request merged**